src/python/pants/backend/python/rules/resolve.py

```diff
@@ -0,0 +1,356 @@
+from pants.engine.target import Dependencies, Targets, TransitiveTargets
+from pants.option.custom_types import GlobExpansionConjunction
+from pants.option.global_options import GlobMatchErrorBehavior
+from pants.python.python_repos import PythonRepos
+from pants.python.python_setup import PythonSetup
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class Resolver(PexTool):
+    _entry_point = "resolver:main"
+    _module_names = ("resolver", "common")
+
+    pex: Pex
+
+
+@named_rule(desc="Create python resolver")
+async def create_resolver(pex_bin: PexBin) -> Resolver:
+    resolver_snapshot = await Get[Snapshot](InputFilesContent, Resolver.files())
+    init_injected = await Get[InitInjectedSnapshot](InjectInitRequest(snapshot=resolver_snapshot))
+    return Resolver(
+        pex=await Get[Pex](
+            PexRequest(
+                output_filename="bin/resolver.pex",
+                requirements=PexRequirements((pex_bin.requirement, "dataclasses==0.6")),
+                sources=init_injected.snapshot.digest,
+                entry_point=Resolver.entry_point(),
+                additional_args=("--no-strip-pex-env", "--unzip"),
+            )
+        )
+    )
+
+
+@dataclass(frozen=True)
+class PexCache:
+    snapshot: Snapshot = EMPTY_SNAPSHOT
+    path: str = "pex_root"
+
+
+@dataclass(frozen=True)
+class ResolveRequest:
+    requirements: PexRequirements
+    cache: PexCache = PexCache()
+    local: bool = False
+    requirer: Optional[str] = None
+
+
+@dataclass(frozen=True)
+class ResolveResult:
+    dependencies: DistributionDependencies = DistributionDependencies()
+    cache: PexCache = PexCache()
+
+
+@named_rule(desc="Resolve python requirements")
+async def resolve_requirements(
+    resolver: Resolver,
+    python_setup: PythonSetup,
+    python_repos: PythonRepos,
+    request: ResolveRequest,
+    subprocess_encoding_environment: SubprocessEncodingEnvironment,
+    log_level: LogLevel,
+) -> ResolveResult:
+    argv = list(PexDebug(log_level).iter_pex_args())
+
+    if python_setup.requirement_constraints is not None:
+        constraint_file_snapshot = await Get[Snapshot](
+            PathGlobs(
+                [python_setup.requirement_constraints],
+                glob_match_error_behavior=GlobMatchErrorBehavior.error,
+                conjunction=GlobExpansionConjunction.all_match,
+                description_of_origin="the option `--python-setup-requirement-constraints`",
+            )
+        )
+        argv.extend(["--constraints", python_setup.requirement_constraints])
+    else:
+        constraint_file_snapshot = EMPTY_SNAPSHOT
+
+    for interpreter_search_path in python_setup.interpreter_search_paths:
+        argv.extend(["--interpreter-search-path", interpreter_search_path])
+
+    for interpreter_constraint in python_setup.interpreter_constraints:
+        argv.extend(["--interpreter-constraint", interpreter_constraint])
```

Comment:
In `pex.py`, we only specify platforms or interpreter constraints, but not both. 

https://github.com/pantsbuild/pants/blob/766ce1d0127a3c59afb6d8cca533616fa427e57d/src/python/pants/backend/python/rules/pex.py#L289-L297

We also don't handle `python_setup` because of not knowing how it should be consumed: https://github.com/pantsbuild/pants/issues/9562

--

It's possible that `pex.py` is wrong. Either way, we probably want consistency.